### PR TITLE
[runtime] Allow valuetypes larger than 1MB created by csc. Fixes #60288.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

File: mono/metadata/class.c [hunk 1]
```diff
@@ -2195,8 +2195,12 @@ mono_class_layout_fields (MonoClass *klass, int base_instance_size, int packing_
 	}
 
 	/*valuetypes can't be neither bigger than 1Mb or empty. */
-	if (klass->valuetype && (klass->instance_size <= 0 || klass->instance_size > (0x100000 + sizeof (MonoObject))))
-		mono_class_set_type_load_failure (klass, "Value type instance size (%d) cannot be zero, negative, or bigger than 1Mb", klass->instance_size);
+	if (klass->valuetype && (klass->instance_size <= 0 || klass->instance_size > (0x100000 + sizeof (MonoObject)))) {
+		/* Special case compiler generated types */
+		/* Hard to check for [CompilerGenerated] here */
+		if (!strstr (klass->name, "StaticArrayInitTypeSize"))
```
Comment (joncham) [inline mono/metadata/class.c]: This seems to be what Roslyn emits. If you are not worried about `mcs` I understand. But if I still need to worry about `mcs`, I should be able to check for `$ArrayType` as well?

File: mono/metadata/class.c [hunk 2]
```diff
@@ -2195,8 +2195,12 @@ mono_class_layout_fields (MonoClass *klass, int base_instance_size, int packing_
 	}
 
 	/*valuetypes can't be neither bigger than 1Mb or empty. */
-	if (klass->valuetype && (klass->instance_size <= 0 || klass->instance_size > (0x100000 + sizeof (MonoObject))))
-		mono_class_set_type_load_failure (klass, "Value type instance size (%d) cannot be zero, negative, or bigger than 1Mb", klass->instance_size);
+	if (klass->valuetype && (klass->instance_size <= 0 || klass->instance_size > (0x100000 + sizeof (MonoObject)))) {
+		/* Special case compiler generated types */
+		/* Hard to check for [CompilerGenerated] here */
+		if (!strstr (klass->name, "StaticArrayInitTypeSize") && !strstr (klass->name, "!ArrayType"))
```
Comment (joncham) [inline mono/metadata/class.c]: Did you intend `$ArrayType`?

**Comment by vargaz** [comment]
build

**Comment by kumpera** [comment]
Test failures are unrelated.

**Pull request merged**